src/lib/math.js

```diff
@@ -293,16 +293,37 @@ var $builtinmodule = function (name) {
         return new Sk.builtin.float_(Math.sqrt((x * x) + (y * y)));
     });
 
+    var MAX_SAFE_INTEGER_FACTORIAL = 18; // 19! > Number.MAX_SAFE_INTEGER
     mod.factorial = new Sk.builtin.func(function (x) {
         Sk.builtin.pyCheckArgs("factorial", arguments, 1, 1);
         Sk.builtin.pyCheckType("x", "number", Sk.builtin.checkNumber(x));
 
         x = Math.floor(Sk.builtin.asnum$(x));
         var r = 1;
-        for (var i = 2; i <= x; i++) {
+        for (var i = 2; i <= x && i <= MAX_SAFE_INTEGER_FACTORIAL; i++) {
             r *= i;
         }
-        return new Sk.builtin.int_(r);
+        if(x<=MAX_SAFE_INTEGER_FACTORIAL){
+            return new Sk.builtin.int_(r);
+        }else{
+            // for big numbers (19 and larger) we first calculate 18! above
+            // and then use bigintegers to continue the process.
+
+            // This is inefficient as hell, but it produces correct answers.
```

Comment:
Would it make sense to create an array to cache these results? You could do something like `cache[x] = result` before returning.

Reply:
I agree that this is appealing on the surface.

But I have several arguments against it, the most important being: "is bignum factorial actually used in any algorithms that would benefit from caching?" Nobody has reported this issue yet, nobody has fixed this issue yet.

I only noticed because I was doing interactive exploring of math with my kids. I wanted to show them how big the number for 52! is. (number of combinations of a deck)

I don't know why I would need to do that in an algorithm that would benefit from caching. And before you start caching you need to guess how big the cache should be lest you accidentally implement a memory leak. We'd only know how big to make the cache if we could guess what algorithm is using it and for what.

Reply:
That's fair, but one could imagine large benefits from even a small cache, if you simply use it to start the computation.  For example, even a 2-entry cache is likely be helpful if you store 18! and the largest number computed so far. If you then use one or the other to begin the computation, you will always at least skip the re-computation of 18!.

Your argument is completely reasonable, though.  I have known about this for a long time and not bothered to do anything about it, because it's just not that common.  I thought I had filed an issue, but looking through the open issues, that does not appear to be the case.

To be clear, I am not opposed to the pull request because of the lack of caching. Regardless of whether caching is added, it fixes a correctness bug so we should definitely pull it in to Skulpt.  (I do think that independent tests for ```bisect``` should be considered, though.)

Reply:
I took a stab at caching. It speeds up this program by about 40%:

start = time()
for i in range(0,10):
    for i in range(0, 1000):
        factorial(i)
print time()-start

To do it "right" seemed more complex than just caching 18! because computing 18! is so fast. It's the big numbers you want to cache, so I cache every power of 2. And you don't want this to be a simple memoization of inputs and outputs. You want to "jump" to a cached value before you start looping.

Ultimately the caching code is kind of overwhelming the factorial code so I'm not sure if I should clean up the code a bit and pull request it or keep the simpler, easier to read factorial-only code. I don't feel strongly either way.

https://github.com/prescod/skulpt/compare/master...prescod:factorial_with_bigints_and_caching?expand=1